research-prototype/Programs/Source/erik.mpc

```diff
@@ -0,0 +1,238 @@
+                                                              size = numrows - 1))
+    stopbit[numrows - 1] = 0
+    credit[numrows - 1] = repval[numrows - 1]
+
+    credit.assign_vector(repval.get_vector(size = numrows - 1)
+                         + stopbit.get_vector(size = numrows - 1)
+                         * repval.get_vector(base = 1,
+                                             size = numrows - 1))
+    
+    zeros = Array(numrows // 2, sint)
+    zeros.assign_all(0)
+
+    stepsize = 1
+
+    while stepsize < numrows // 2:
+
+        # inc(size, base, step, repeat)
+        # produces a regint vector like range(base, size // step, step)
+        # get method takes an array of regints and produces the
+        # vector indexed by them,
+
+        # flag = helperbits_and_istrigger{stepsize:]
+        # credit = reports[: -stepsize][VALUE] + flag * reports[stepsize: ][VALUE]
+        stepsize *= 2
+
+        new_size = numrows - stepsize
+
+        flag = (stopbit.get_vector(size = new_size)
+                * helperbits_and_istrigger.get_vector(base = stepsize,
+                                                     size = new_size))
+        new_credit = (credit.get_vector(size = new_size)
+                      + flag
+                      * credit.get_vector(base = stepsize,
+                                          size = new_size))
+        stopbit.assign_vector(flag
+                              * stopbit.get_vector(base = stepsize,
+                                                   size = new_size))
+        # Now 0 out the last stepsize stopbit's
+        # stopbit.get_vector(base = new_size, size = stepsize).assign_all(0)
+        stopbit.assign_vector(zeros.get_vector(size = stepsize),
+                              base = new_size)
+            
+        credit.assign_vector(new_credit)
+
+    # Calculate final_credits
+    if breakdown_values:
+
+        breakdown = reports.get_column(BREAKDOWN_KEY)
+
+        # Once we have AND, XOR is just addition and subtraction
+
+        helperbits_xor_istrigger = (helperbits
+                                    + is_trigger.get_vector()
+                                    - helperbits_and_istrigger)
+
+        final_credits = helperbits_xor_istrigger * credit
+        
+        breakdown_keys = list(range(breakdown_values))
+
+        breakdown_key_sums = Array(breakdown_values, sint)
+
+        for breakdown_key in breakdown_keys:
+
+            # max[idx] = 1 if breakdown[idx] == breakdown_key, 0 otherwise
+            mask = breakdown == breakdown_key
+            # Sum up the final_credits for this breakdown
+            # We can do this with sum, but tree_reduce is more efficient.
+            breakdown_key_sums[breakdown_key] = tree_reduce(lambda x,y: x + y,
+                                                            mask * final_credits)
+            return breakdown_key_sums
+
+        return None
+
+def main():
```

Comment:
not sure if this is particularly helpful? it just forces everything to be indented and doesn't actually change the functionality